File: frontend/src/application/use-cases/user-use-case.ts
/**
 * User use cases - application logic for user operations.
 */

import { User } from '@/src/domain/entities';
import {
  UserNotFoundException,
  ValidationException,
} from '@/src/domain/exceptions';
import { IUserRepository } from '@/src/domain/interfaces/user-repository.interface';
import { UpdateUserData } from '@/src/domain/interfaces/user-repository.interface';

/**
 * User use cases.
 */
export class UserUseCase {
  constructor(private readonly userRepository: IUserRepository) {}

  /**
   * Get current authenticated user.
   *
   * @returns Current user
   * @throws {UserNotFoundException} If user not found
   */
  async getCurrentUser(): Promise<User> {
    try {
      return await this.userRepository.getCurrentUser();
    } catch (error) {
      if (error instanceof UserNotFoundException) {
        throw error;
      }
      throw new UserNotFoundException('Current user not found');
    }
  }

  /**
   * Get user by ID.
   *
   * @param userId - User ID
   * @returns User entity
   * @throws {UserNotFoundException} If user not found
   */
  async getUserById(userId: string): Promise<User> {
    try {
      return await this.userRepository.getById(userId);
    } catch (error) {
      if (error instanceof UserNotFoundException) {
        throw error;
      }
      throw new UserNotFoundException(`User with ID ${userId} not found`);
    }
  }

  /**
   * Get all users.
   *
   * @returns List of all users
   */
  async getAllUsers(): Promise<User[]> {
    return this.userRepository.getAll();
  }

  /**
   * Update user profile.
   *
   * @param userId - User ID
   * @param data - Update data
   * @returns Updated user
   * @throws {UserNotFoundException} If user not found
   * @throws {ValidationException} If update data is invalid
   */
  async updateUser(
    userId: string,
    data: UpdateUserData
  ): Promise<User> {
    try {
      return await this.userRepository.update(userId, data);
    } catch (error) {
      if (
        error instanceof UserNotFoundException ||
        error instanceof ValidationException
      ) {
        throw error;
      }
      throw new ValidationException(
        `Failed to update user: ${error instanceof Error ? error.message : 'Unknown error'}`
      );
    }
  }

  /**
   * Check if user exists.
   *
   * @param userId - User ID
   * @returns True if user exists, false otherwise
   */
  async userExists(userId: string): Promise<boolean> {
    return this.userRepository.exists(userId);
  }
}

File: backend/api/application/use_cases/user_use_case.py
"""User use cases - application logic for user operations."""

from typing import List, Optional

from api.application.interfaces.user_repository import IUserRepository
from api.domain.exceptions import UserNotFoundException
from api.models import User


class UserUseCase:
    """Use cases for user operations."""

    def __init__(self, user_repository: IUserRepository) -> None:
        """
        Initialize user use case.

        Args:
            user_repository: Repository for user data access
        """
        self._user_repository = user_repository

    def get_user_by_id(self, user_id: int) -> User:
        """
        Get user by ID.

        Args:
            user_id: User ID

        Returns:
            User entity

        Raises:
            UserNotFoundException: If user not found
        """
        user = self._user_repository.get_by_id(user_id)
        if not user:
            raise UserNotFoundException(f"User with ID {user_id} not found")
        return user

    def get_user_by_email(self, email: str) -> Optional[User]:
        """
        Get user by email.

        Args:
            email: User email

        Returns:
            User entity or None if not found
        """
        return self._user_repository.get_by_email(email)

    def get_all_users(self) -> List[User]:
        """
        Get all users.

        Returns:
            List of all users
        """
        return self._user_repository.get_all()

    def update_user(self, user: User) -> User:
        """
        Update user.

        Args:
            user: User to update

        Returns:
            Updated user

        Raises:
            UserNotFoundException: If user not found
        """
        if not self._user_repository.exists(user.id):
            raise UserNotFoundException(f"User with ID {user.id} not found")
        return self._user_repository.save(user)

